doc/manage-duniverse.mld

```diff
@@ -0,0 +1,116 @@
+Calling [opam monorepo pull] will create a [duniverse] folder into which it
+will unpack the sources of the packages that are mentioned in the lockfile.
+
+{2 Not locked}
+
+While this is not a [opam-monorepo] workflow per se, projects that do not use
+[opam-monorepo] can still be used with [opam-monorepo]. As such every
+contributor is responsible for creating and maintianing their own lockfile and
+duniverse.
+
+Advantages include:
+
+{ul
+{- No opt-in to [opam-monorepo] required, every developer can chose to use it
+   or not}
+}
+
+Unfortunately, in this approach the disadvantages dominate:
+
+{ul
+{- Everyone might end up with different packages in their duniverse, thus low
+   reproducability}
+{- Everyone needs to have opam-monorepo installed}
+}
+
+The [opam-monorepo] maintainers don't recommend this approach, as is has hardly
+any advantages over an [opam-monorepo]-less approach.
+
+{2:lockfile-in-git Lockfile in Git}
+
+Another approach is very similar to the regular opam workflow, but in addition
+to having the [opam] file committed to the repository, also the lockfile that
+[opam monorepo lock] generates is added.
+
+Advantages include:
+
+{ul
+{- The git repostitory does not grow by much, only one file}
+{- The packages to be installed are reproducible}
+{- Installation on the packages regularly into the Opam switch is still
+   possible}
+}
+
+Disadvantages might be
+
+{ul
+{- Everyone needs to have opam-monorepo installed to pull the dependencies}
+}
+
+[opam-monorepo] maintainers encourage people to adopt this pattern, since it
+provides a good practical compromise between reproducibility and
+practicability.
+
+{2:duniverse-in-git Duniverse in Git}
+
+The most extreme approach is to create a lockfile with [opam monorepo lock],
+download all the dependencies via [opam monorepo pull] and then commit both the
+lockfile as well as the pulled sources.
+
+This appraoch has some advantages:
+
+{ul
+{- Only the person that creates the lockfile and duniverse must have
+   [opam-monorepo] installed}
+{- The repository can be immediately used after cloning as everything is
+   self-contained}
+{- Can be used offline}
+}
+
+Yet unfortunately it also has some disadvantages:
+
+{ul
+{- Contributors unaware of [opam-monorepo] might wonder where the sources
+   come from}
+{- Can lead to lots of files in repository}
+{- Potentially complicated merge conflicts if multiple contributors update
+   versions of locked packages}
+{- Users might modify the committed sources, yet those patches will be
+   overwritten when [opam monorepo pull] is run, leading to changes
+   disappearing}
+{- Imported packages can interact with version control in unexpected ways,
+   see {!label-git-add}}
+}
+
+For ease of collaboration and to keep the repositories manageable
+[opam-monorepo] maintainers recommend against this approach and suggest the
+approach in {!label-lockfile-in-git} instead.
```

Comment:
I do not agree: there are valid use-cases for committing vendor code to the repositories, let's not try to dictate their requirements to our users. Would be much better to just reckon this, and point to possible issues and fixes (as the next paragraph does well).

Reply:
Indeed, irrespective of the recommendations of the opam-monorepo maintainers, I am going to continue checking in my code into duniverse/ :-)  The reason for this is that the benefit of this approach is somewhat understated and hidden away:
> The repository can be immediately used after cloning as everything is self-contained

The above is an absolutely massive benefit for newcomers to a project:
- it doesn't require opam to be installed, and, more crucially, doesn't require a working `opam init` to have happened.
- only `dune` is required for a build, and (since dune 2.0) the `dune` binary can be distributed independently of the dune libraries (which are vendored).   Thus a newcomer can `apt install ocaml-dune` and get on with things.
- it doesn't require opam-monorepo to remain working into the future.  I can `git checkout` an older revision of RWO and have a reasonable expectation that it'll build, for example.

This guide is an excellent start, but it is organised around "operational" uses of monorepo. But a reader of the document unfamiliar with opam-monorepo is going to start from a different mindset: they are going to wonder how best to serve various users of their projects and achieve outcomes such as "distribute this binary for people that don't know ocaml", to "make first contribution to the core tool", to "experienced user wants to patch a library and upstream the change".

If you organise things by "how deeply someone wants to get into the project", then the ordering of how you introduce vendoring, lockfiles and on-the-fly locking changes.